### PR TITLE
Proposal: deprecate autovalidate parameter of the Form, FormField and TextFormField widget

### PR DESCRIPTION
**Pull request by pedromassango**
## Description

This PR comes to deprecate the `autovalidate` parameter of the `Form`, `FormField` and the `TextFormField` widget in favor of the recently added `autovalidateMode` parameter. This is a follow up of #59766.

## Related Issues

This does not need tests since I'm only deprecating a parameter.

## Tests

I added the following tests:

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

**Comment by pedromassango** [comment]
@chunhtai this is the PR to deprecate the `autovalidate` parameter as we talked on #59766.

**Comment by chunhtai** [comment]
it looks like the test in https://github.com/jocosocial/rainbowmonkey is failing, you should be able to repro using https://github.com/flutter/tests/blob/master/registry/rainbowmonkey.test

you can file an issue or pr to fix it in that repo, cc @Hixie for the repo owner.

**Comment by chunhtai** [comment]
> Hi @chunhtai the PR was merged on https://github.com/jocosocial/rainbowmonkey but the tests here are not passing. I think this has something to do with the flutter build or rainbowmonkey's build.
> 
> Also Hixie pointed something here [jocosocial/rainbowmonkey#329 (comment)](https://github.com/jocosocial/rainbowmonkey/pull/329#issuecomment-660542629) that may cause this.

You will need to update the flutter/test library to point to the new ref commit.
https://github.com/flutter/tests/blob/master/registry/rainbowmonkey.test

**Comment by chunhtai** [comment]
https://github.com/flutter/tests/pull/43 has merged, please do a rerun or rebased and the tests should pass

**Comment by pedromassango** [comment]
@chunhtai Done.

**Comment by chunhtai** [comment]
looks like there are some more internal or external test that need to be migrate

**Comment by pedromassango** [comment]
@chunhtai looks like another PR just landed in the Gallery repository and I think that the last commit there has my changes, is there a way to confirm that or restart this PR tests?

**Comment by chunhtai** [comment]
> @chunhtai looks like another PR just landed in the Gallery repository and I think that the last commit there has my changes, is there a way to confirm that or restart this PR tests?

I just clicked rerun, you can also trigger rerun by commit a new commit (ex, rebase your branch)

**Comment by chunhtai** [comment]
Hi @pedromassango cirrus seems not happy

**Comment by justinmc** [comment]
If merging in master isn't working, it might need to have all the commits squashed, or rebase master.

**Comment by pedromassango** [comment]
> If merging in master isn't working, it might need to have all the commits squashed, or rebase master. The failure I looked at was unrelated to the changes.

Not sure why tests are failing too. I will `rebase master` and see how it goes.

**Comment by pedromassango** [comment]
@justinmc I probably need some help with this, what else I can do to make the tests to pass?

**Comment by justinmc** [comment]
@pedromassango If you squash all of your commits into 1, then rebase off of master, then force push, that should definitely update everything.  Sorry that that will make you lose your history, though.  I think the idea is that the first commit's SHA needs to change in order for something to get updated and allow the PR to pass.

I see a docs failure that might be real, but I also see some version failures that don't make sense and probably do need the steps above to get fixed.

**Comment by chunhtai** [comment]
This file may require migration as well
```
dev/integration_tests/flutter_gallery/lib/demo/material/text_form_field_demo.dart:183:11 • deprecated_member_use
```

**Comment by chunhtai** [comment]
there are also doc string error
```
dartdoc:stderr:   error: unresolved doc reference [autovalidate]
dartdoc:stderr:     from widgets.Form.autovalidateMode: (file:///b/s/w/ir/k/flutter/packages/flutter/lib/src/widgets/form.dart:144:26)
dartdoc:stderr:     in documentation inherited from widgets.Form.autovalidateMode: (file:///b/s/w/ir/k/flutter/packages/flutter/lib/src/widgets/form.dart:144:26)
dartdoc:stderr:   error: unresolved doc reference [autovalidate]
dartdoc:stderr:     from widgets.FormField.autovalidateMode: (file:///b/s/w/ir/k/flutter/packages/flutter/lib/src/widgets/form.dart:388:26)
dartdoc:stderr:     in documentation inherited from widgets.FormField.autovalidateMode: (file:///b/s/w/ir/k/flutter/packages/flutter/lib/src/widgets/form.dart:388:26)
```

File: packages/flutter/lib/src/widgets/form.dart [hunk 1]
```diff
@@ -381,8 +381,8 @@ class FormField<T> extends StatefulWidget {
   /// will auto validate even without user interaction and
   /// if [AutovalidateMode.disabled] the auto validation will be disabled.
   ///
-  /// Defaults to [AutovalidateMode.disabled] if [autovalidate] is false which
-  /// means no auto validation will occur. If [autovalidate] is true then this
+  /// Defaults to [AutovalidateMode.disabled] if autovalidate is false which
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/form.dart:384]: you can use `autovalidate`

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/form.dart [hunk 2]
```diff
@@ -114,11 +119,6 @@ class Form extends StatefulWidget {
   /// {@macro flutter.widgets.child}
   final Widget child;
 
-  /// If true, form fields will validate and update their error text
-  /// immediately after every change. Otherwise, you must call
-  /// [FormState.validate] to validate.
-  final bool autovalidate;
```
Comment (skyeskie) [inline packages/flutter/lib/src/widgets/form.dart:120]: Turns out this is a breaking change. The public field is removed, not deprecated.

This affects (at least) [flutter_form_builder](https://pub.dev/packages/flutter_form_builder).
https://github.com/danvick/flutter_form_builder/issues/465

Ideally, should (with appropriate deprecated tag) put something like
```dart
bool get autovalidate => autovalidateMode != AutovalidateMode.disabled;
```

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/form.dart:120]: @pedromassango oops did we miss this one?

Comment (pedromassango) [reply packages/flutter/lib/src/widgets/form.dart:120]: Hmm. Just checked `form.dart` file and I don't see any reference for `final bool autovalidate`. We did removed this intentionally in favour of AutovalidateMode property.

Maybe this is referencing a old commit file?!

Comment (pedromassango) [reply packages/flutter/lib/src/widgets/form.dart:120]: > bool get autovalidate => autovalidateMode != AutovalidateMode.disabled;

This is interesting and would prevent the breaking changes 🤔 
@chunhtai do you think this we should put it back?

Comment (pedromassango) [reply packages/flutter/lib/src/widgets/form.dart:120]: @skyeskie
What are the use cases of using this Widget's property?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/form.dart:120]: I think we should put this one back, but we should do
```
autovalidateMode = autovalidateMode ?? autovalidate ? AutovalidateMode.always : AutovalidateMode.disabled
```
Same as what we do for other formfield

Comment (pedromassango) [reply packages/flutter/lib/src/widgets/form.dart:120]: Alright, I will send a PR to put it back ASAP

Comment (skyeskie) [reply packages/flutter/lib/src/widgets/form.dart:120]: `FormField` has the exact same issue (below, 363-368)

@pedromassango
I personally don't use this, but just ran across when `flutter_form_builder` refused to compile, specifically when it tries to access `autovalidate` on `FormField` in its `FormBuilderCustomField` class [here](https://github.com/danvick/flutter_form_builder/blob/2d5f61b3863b058b9b6213aac25f6b62b65bd44e/lib/src/form_builder_custom_field.dart#L96).

Comment (pedromassango) [reply packages/flutter/lib/src/widgets/form.dart:120]: The PR is: https://github.com/flutter/flutter/pull/66267

@chunhtai I think we should add it back on formField widget too, right?

**Pull request merged**